cockatrice/src/client/ui/layouts/flow_layout.cpp

```diff
@@ -0,0 +1,293 @@
+            currentYPosition += rowHeight + verticalSpacing();
+            rowWidth = 0;
+            rowHeight = 0;
+        }
+        rowItems.append(item);
+        rowWidth += itemSize.width() + horizontalSpacing();
+        rowHeight = qMax(rowHeight, itemSize.height());
+        currentXPosition += itemSize.width() + horizontalSpacing();
+    }
+
+    layoutRow(rowItems, originX, currentYPosition);
+
+    return currentYPosition + rowHeight;
+}
+
+/**
+ * @brief Helper function for arranging a single row of items within specified bounds.
+ * @param rowItems Items to be arranged in the row.
+ * @param x The x-coordinate for starting the row.
+ * @param y The y-coordinate for starting the row.
+ */
+void FlowLayout::layoutRow(const QVector<QLayoutItem *> &rowItems, int x, int y)
+{
+    for (QLayoutItem *item : rowItems) {
+        QSize itemMaxSize = item->widget()->maximumSize();
+        int itemWidth = qMin(item->sizeHint().width(), itemMaxSize.width());
+        int itemHeight = qMin(item->sizeHint().height(), itemMaxSize.height());
+        item->setGeometry(QRect(QPoint(x, y), QSize(itemWidth, itemHeight)));
+        x += itemWidth + horizontalSpacing();
+    }
+}
+
+/**
+ * @brief Returns the preferred size for this layout.
+ * @return The maximum of all item size hints as a QSize.
+ */
+QSize FlowLayout::sizeHint() const
+{
+    QSize size;
+    for (QLayoutItem *item : items) {
+        size = size.expandedTo(item->sizeHint());
+    }
+    return size.isValid() ? size : QSize(0, 0);
+}
+
+/**
+ * @brief Returns the minimum size required to display all layout items.
+ * @return The minimum QSize needed by the layout.
+ */
+QSize FlowLayout::minimumSize() const
+{
+    QSize size;
+    for (QLayoutItem *item : items) {
+        size = size.expandedTo(item->minimumSize());
+    }
+
+    size.setWidth(qMin(size.width(), getParentScrollAreaWidth()));
+    size.setHeight(qMin(size.height(), getParentScrollAreaHeight()));
+
+    return size.isValid() ? size : QSize(0, 0);
+}
+
+/**
+ * @brief Adds a new item to the layout.
+ * @param item The layout item to add.
+ */
+void FlowLayout::addItem(QLayoutItem *item)
+{
+    items.append(item);
+}
+
+/**
+ * @brief Retrieves the count of items in the layout.
+ * @return The number of layout items.
+ */
+int FlowLayout::count() const
+{
+    return items.size();
+}
+
+/**
+ * @brief Returns the layout item at the specified index.
+ * @param index The index of the item to retrieve.
+ * @return A pointer to the item at the specified index, or nullptr if out of range.
+ */
+QLayoutItem *FlowLayout::itemAt(int index) const
+{
+    return items.value(index);
+}
+
+/**
+ * @brief Removes and returns the item at the specified index.
+ * @param index The index of the item to remove.
+ * @return A pointer to the removed item, or nullptr if out of range.
+ */
+QLayoutItem *FlowLayout::takeAt(int index)
+{
+    return (index >= 0 && index < items.size()) ? items.takeAt(index) : nullptr;
```

Comment:
What's the point of boundary checking if the itemAt returns regardless of index?